File: content/Structures/DSU.cpp
// DSU
//
struct DSU
{
    vector<int>par, sz;
    int comp;
    DSU(int n)
    {
        par.resize(n);
        comp = n;
        sz = vector<int>(n, 1);
        iota(par.begin(), par.end(), 0);
    }
    int find(int x)
    {
        if (x != par[x])
            return par[x] = find(par[x]);
        return x;
    }
    bool merge(int a, int b)
    {
        a = find(a);
        b = find(b);
        if (a == b)
            return false;
        if (sz[a] < sz[b])
            swap(a, b);
        comp--;
        par[b] = a;
        sz[a] += sz[b];
        sz[b] = 0;
        return true;
    }
    int get_sz(int a) {
        return sz[find(a)];
    }
};

File: Kotlin/Structures/DSU.kt
class DSU(N_:Int) {
    val par:Array<Int>
    val sz:Array<Int>
    var comp:Int
    init {
    	comp = N_
        par = Array<Int>(N_){it}
        sz = Array<Int>(N_){_ -> 1}
    }
    fun find(x:Int):Int {
        if (x != par[x]) {
            par[x] = find(par[x])
            return par[x]
        }
        return x
    }
    fun merge(a_:Int, b_:Int):Boolean {
    	var a = a_
    	var b = b_
        a = find(a); b = find(b);
        if (a == b) {
            return false
        }
        if (sz[a] < sz[b]) {
            a = b.also{b = a}
        }
        comp--
        par[b] = a
        sz[a] += sz[b]
        sz[b] = 0
        return true
    }
    fun get_sz(a:Int):Int {
        return sz[find(a)]
    }
}